server/src/arcs-master-app.ts

```diff
@@ -32,41 +40,128 @@ class ArcsMasterApp extends AppBase {
    */
   gcpSafeIdentifier(str) {
-    return str.replace(/[ +-=]/g, '').toLowerCase().substring(0, 32);
+    return str.replace(/[ +\-=]/g, '').toLowerCase().substring(0, 32);
   }
 
-  async findDeployment(req, res, next) {
+  async lock(req, res, next) {
+    console.log("fingerprint is " + req.params.fingerprint);
     const fingerprint = this.gcpSafeIdentifier(req.params.fingerprint);
+    console.log("gcp safe fingerprint " + fingerprint);
+    try {
+      const cloud = CloudManager.forGCP();
+      const disk = await cloud.disks().find(fingerprint);
+      if (disk) {
+        const attached = await disk.isAttached();
+
+        if (attached) {
+          const container: Container | null = await cloud.containers().find(fingerprint);
+          if (container != null) {
+            await fetch("http://" + container.url() + "/lock");
+            disk.dismount();
+            return;
+          }
 
+        }
+      }
+    } catch (e) {
+      console.log(e);
+    }
+  }
+
+  async unlock(req, res, next) {
+    console.log("fingerprint is " + req.params.fingerprint);
+    const fingerprint = this.gcpSafeIdentifier(req.params.fingerprint);
+    console.log("gcp safe fingerprint " + fingerprint);
     try {
       const cloud = CloudManager.forGCP();
-      const container: Container | null = await cloud.containers().find(fingerprint);
-      if (container != null) {
-        const url = container.url();
-        if (url === 'pending') {
-          res.send('Creation still in progress, refresh for status.');
+      const disk = await cloud.disks().find(fingerprint);
+      if (disk) {
+        const attached = await disk.isAttached();
+        if (attached) {
+          return;
+        }
+        const container: Container | null = await cloud.containers().find(fingerprint);
+        if (container != null) {
+          await disk.mount(req.params.rewrappedKey, await container.node());
+          await fetch("http://" + container.url() + "/unlock");
+          return;
         } else {
-          res.send(url);
+          // TODO: if container not found, deploy a new one?
         }
-      } else {
-        res.send('Container ' + fingerprint + ' not found.');
       }
+    } catch (e) {
+      console.log(e);
+    }
+  }
 
+  async findDeployment(req, res, next) {
+    console.log("fingerprint is " + req.params.fingerprint);
+    const fingerprint = this.gcpSafeIdentifier(req.params.fingerprint);
+    console.log("gcp safe fingerprint " + fingerprint);
+    try {
+      const cloud = CloudManager.forGCP();
+      const disk = await cloud.disks().find(fingerprint);
+      if (disk) {
+        const attached = await disk.isAttached();
+        const wrappedKey  = await disk.wrappedKeyFor(fingerprint);
+
+        if (attached) {
+          const container: Container | null = await cloud.containers().find(fingerprint);
+          if (container != null) {
+            const status = container.status();
+            if (status !== 'Running') {
```

Comment:
Here and below is a bit hard to follow due to deep nesting and combination of `else` and `return`, could maybe restructure a bit with helper methods or using either `else` or `return` rather than both, could do later, just mentioning while reading.

Reply:
I restructed with early-exits